Import csv

Nisha-Nathan opened the pull request:
This PR request includes a POST request that receives a csv file that is parsed with a code written to parse csv files with certain attributes and then add the csv file to a corpus

ADJohnson314 committed:
Spiked script parsed csv

Edited code and created new text files for TED Talks

ADJohnson314 committed:
Created auto_instances.py

ADJohnson314 committed:
Fixed csv to txt code & improved doc to model instance code

With this code, all rows of the csv are now successfully converted into txt files, regardless of the characters in the title column. In addition, the code in auto_instances now uses the Document manager to easily and efficiently create new instances of the Document model.

ADJohnson314 committed:
Added Corpus creation code

Wrote new code that adds the Document instances for each TED Talk file into a Corpus object (an instance of the Corpus model).

ADJohnson314 committed:
Streamlined process for turning csv data into a Corpus

The program no longer needs to turn reach row in the csv into its own text file. Each row of data in the csv is now directly converted into a Document, and the Documents are all combined into a Corpus.

Nisha-Nathan committed:
POST request

 Created a POST request that took in a csv file and returned a corpus(a serialized version of a corpus)

ADJohnson314 committed:
Added POST request functionality

Wrote an API endpoint function that receives through a POST request a filename for a file in the backend (i.e. small_talks.csv). It then runs the corresponding file through parse_csv and returns as a Response a text representation of the corpus created from that file.

Nisha-Nathan committed:
Delete views.py

Nisha-Nathan committed:
Delete settings.py

Nisha-Nathan committed:
Delete manage.py

Nisha-Nathan committed:
Delete __init__.py

Nisha-Nathan committed:
Delete asgi.py

Nisha-Nathan committed:
Delete urls.py

Nisha-Nathan committed:
Delete admin.py

ADJohnson314 committed:
Merge branch 'import_csv' of https://github.com/dhmit/gender_analysis_web into import_csv

ADJohnson314 committed:
Merge branch 'import_csv' of https://github.com/dhmit/gender_analysis_web into import_csv

MBJean commented on backend/app/POST.py:
```diff
@@ -0,0 +1,18 @@
+    x= parse_csv(csv_reader)#returns None because parse_csv doesn't return anything
+    return x
+
+
```
I think we talked about this file being some spiked, testing work. Do I remember that correctly? If so, we can probably delete it now that we've incorporated the work into `views.py`.

MBJean commented on backend/config/urls.py:
```diff
@@ -43,5 +45,6 @@
     path('corpora', views.corpora, name='corpora'),
-    path('corpus/<int:corpus_id>', views.corpus, name='corpus')
+    path('corpus/<int:corpus_id>', views.corpus, name='corpus'),
+
```
Looks like we've got an extraneous space here that we can delete.

MBJean commented on backend/config/urls.py:
```diff
@@ -35,6 +35,8 @@
     path('api/delete_corpus', views.delete_corpus),
     path('api/corpus/<int:corpus_id>', views.get_corpus),
+    path('api/corpus_from_csv/', views.create_corpus_csv),
+    path('api/upload_document', views.upload_document),
```
Are both of these API endpoints in use? I _think_ I can see from the `views.py` file that `upload_document` is the endpoint we've settled on. If so, we can delete the second endpoint from this file.

MBJean commented on backend/app/views.py:
```diff
@@ -211,6 +228,16 @@ def update_corpus_docs(request):
+    filename = request.data["csv"]
+    return_corpus = create_instances_newer(filename)
+    serializer = CorpusSerializer(return_corpus)
+    return Response(serializer.data)
```
Per my last comment, I think we can remove this function.

MBJean commented on backend/app/views.py:
```diff
@@ -129,6 +133,19 @@ def all_documents(request):
+    API endpoint for uploading csv files that are to be converted to an instance of csv_reader
+    """
+
+    file = request.data["filename"]
```
This is a nitpick, but I'd recommend changing `"filename"` to `"file"`, or something similar. In this case, we're expecting an actual file included in the web request rather than, say, a filename that we'll look up in the server's filesystem. It'll help those of us who work on this project later to have the naming conventions match the data expectations.

MBJean commented on backend/app/views.py:
```diff
@@ -36,7 +38,8 @@
     CorpusSerializer
 )
-
+from app.analysis.auto_instances import create_instances_newer
```
If I'm correct that we can delete `create_corpus_csv` below, we can delete this import, right?

MBJean commented on backend/app/analysis/auto_instances.py:
```diff
@@ -0,0 +1,46 @@
+        return parse_csv(csv_reader, csv_filename)
+
+def main():
+    create_instances_newer("small_talks.csv")
```
I think this is the spiked work that led us to `parse_csv.py` below. If that's the case, can we delete this?

MBJean commented on backend/app/analysis/test.py:
```diff
@@ -0,0 +1,49 @@
+
+
+if __name__ == '__main__':
+    parse_csv(test_csv)
```
I think this is the spiked work that led us to `parse_csv.py` below. If that's the case, can we delete this?

MBJean commented on .idea/temp.iml:
```diff
@@ -17,7 +17,7 @@
       <excludeFolder url="file://$MODULE_DIR$/venv" />
     </content>
-    <orderEntry type="jdk" jdkName="Python 3.9 (gender_analysis_web)" jdkType="Python SDK" />
+    <orderEntry type="inheritedJdk" />
```
This file and `.idea/misc.xml` are autogenerated configuration files for your IDE (PyCharm). This change doesn't look super intentional or meaningful. Mind copying out the changed value and restoring it to what came before?

MBJean commented on .idea/misc.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.9 (gender_analysis_web)" project-jdk-type="Python SDK" />
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.9 (venv) (2)" project-jdk-type="Python SDK" />
```
This file and `.idea/temp.xml` are autogenerated configuration files for your IDE (PyCharm). This change doesn't look super intentional or meaningful. Mind copying out the changed value and restoring it to what came before?

MBJean commented on backend/app/views.py:
```diff
@@ -102,6 +105,7 @@ def add_document(request):
     """
     API endpoint for adding a piece of document
     """
+    print("we are in add_document")
```
Looks like a debugging `print` made it into the PR. We can delete this.

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+    doc_list = []
+    for data in csv_reader:
+        new_doc = None
+        if Document.objects.filter(title=data['title']).count() == 0:  # If the data isn't already an instance in the Document model:
```
Awesome job trying to anticipate the possibility of duplicate records in the database, but it's probably worth thinking more about. That is: filtering `Documents` based on title to find out if the document under iteration (here described as `data`) already exists in the database. It's possible that two separate `Document` instances might have the same title but not be the same text. This is kind of a fundamental problem: what makes a texts unique? Title? Title and author? Title and author and wordcount? That all words match? 

That seems like kind of a big deal to sort out, and beyond the scope of this feature add. I'd recommend we simply dispense with the conditional check here and just write a new `Document`. That is: let later us figure out how to answer "what makes a text unique?"

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+    for data in csv_reader:
+        new_doc = None
+        if Document.objects.filter(title=data['title']).count() == 0:  # If the data isn't already an instance in the Document model:
+            new_doc = Document.objects.create_document(title=data['title'], text=data['transcript'], author=data['speaker_1'], year=data['recorded_date'][:4])
```
I think that this is perfectly suitable for right now, since we're working with a prescribed set of test data (coming from Twitter). But ultimately, we'll likely want to allow the user to include the keys that we'll use to populate these properties. That is: the input dataset might use "publication_date" or something instead of "recorded_date", and we'll need to accommodate that possibility. For now, I'll recommend simply leaving a comment indicating that these keys match the expected Twitter dataset.

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+        doc_list.append(new_doc)
+
+    # Use the id as a unique indentifier instead of the title (or maybe remove this check entirely)
+    if Corpus.objects.filter(title=f"{title} Corpus").count() == 0:  # If the corpus doesn't exist...
```
See my comment above about 'what makes a text unique'. Our `Corpus` model doesn't have a uniqueness constraint on its `title` field, so we may want to simply ignore the problem of checking for duplicates for now.

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+    new_corpus.documents.set(doc_list)
+    new_corpus.save()
+    print(Corpus.objects.all())
+    print("Documents: ")
```
We can remove these debugging print statements.

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+    print(Corpus.objects.all())
+    print("Documents: ")
+    for doc in new_corpus:
+        print(doc.title)
```
And this iteration.

MBJean commented on backend/app/services/parse_csv.py:
```diff
@@ -0,0 +1,37 @@
+    #for doc in tt_corpus:
+       # print(doc.title)
+
+    return tt_corpus
```
It looks like everything past the `return new_corpus` is out of date code, if I had to guess. Let's remove it.

MBJean reviewed (changes requested):
Awesome work! You've tackled numerous parts of the stack and have greatly expanded the input capabilities of the project. I've left a number of comments above, a few of which will need to be addressed before we can merge this. If possible, I'd also love to see a test around `parse_csv`. Great work!

ADJohnson314 committed:
Code revisions

Cleaned up code, removed debugging statements, and deleted unnecessary files, in preparation of merge into main branch.
